parser/src/matrix.rs

```diff
@@ -126,7 +175,125 @@ impl<T: Clone> Matrix<T> {
         }
     }
 }
+
+impl Matrix<Value> {
+    /// Calculates the dot product of two matrices (treated as vectors).
+    ///
+    /// # Errors
+    /// Returns an `Err` if:
+    /// - One of the matricies isn't a vector.
+    /// - The vectors do not have the same size.
+    pub fn dot_product(
+        &self,
+        other: &Matrix<Value>,
+    ) -> Result<Value, EvalError> {
+        // Validation
+        fn vector_err(m: &Matrix<Value>) -> EvalError {
+            return EvalError::IncompatibleMatrixSizes {
+                source: IncompatibleMatrixSizes::Vector {
+                    rows: m.row_count(),
+                    columns: m.column_count(),
+                },
+            };
+        }
+
+        if !self.is_vector() {
+            return Err(vector_err(self));
+        }
+        if !other.is_vector() {
+            return Err(vector_err(other));
+        }
+        if self.get_vector_size() != other.get_vector_size() {
+            return Err(EvalError::IncompatibleMatrixSizes {
+                source: IncompatibleMatrixSizes::SameSizeVectors {
+                    a: self.get_vector_size(),
+                    b: other.get_vector_size(),
+                },
+            });
+        }
+
+        // Calculation
+        let mut sum = Option::None;
+        for i in 0..self.get_vector_size() {
+            let a_i = self.get_vector_element(i);
+            let b_i = other.get_vector_element(i);
+            let term = a_i.mul(&MulType::Implicit, b_i)?;
+            sum = Some(match sum {
+                Some(sum) => (sum + term)?,
+                None => term,
+            });
+        }
+        Ok(sum.expect("Empty vector"))
```

Comment:
This needs a ## Panics section in docs

Reply:
This should never panic because the ::new function ensures the matrix is never empty.

Reply:
Yea but I think its just good praxis to state all assumptions atleast in docs

Reply:
The assumption is documented on the `::new` function. The caller of the `dot_product` function does not have to worry about the size of the matrix, since it will never be empty, so I don't think we should document it here. It's not an assumption for the `dot_product` function. It's sort of like saying "will panic if the matrix is corrupted"

Reply:
That is true I'm wrong